src/core/reference/include/openvino/reference/multinomial.hpp

```diff
@@ -0,0 +1,176 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cmath>
+#include <cstddef>
+
+#include "openvino/reference/broadcast.hpp"
+#include "openvino/reference/convert.hpp"
+#include "openvino/reference/copy.hpp"
+#include "openvino/reference/cum_sum.hpp"
+#include "openvino/reference/divide.hpp"
+#include "openvino/reference/exp.hpp"
+#include "openvino/reference/random_uniform.hpp"
+#include "openvino/reference/slice.hpp"
+
+namespace ov {
+namespace reference {
+namespace multinomial {
+
+template <typename T>
+void test(const T* data, const Shape& shape, int* test_nr) {
+    std::cout << test_nr[0] << "\n"; test_nr[0]++;
+    for(size_t i = 0; i < shape_size<Shape>(shape); ++i) {
+        std::cout << data[i] << " ";
+    } std::cout << " \n";
+}
+/**
+ * @brief Multinomial operation creates a sequence of indices of classes sampled from the multinomial distribution.
+ *
+ * @tparam T Data type of the probs' values.
+ * @tparam U Data type of num_samples' values.
+ * @tparam V Data type of output's values.
+ * @param probs Input tensor containing at each index poisition probability/log probability of sampling a given class.
+ * @param probs_shape Shape of the 'probs' tensor.
+ * @param num_samples Scalar or 1D tensor with a single value that determines the number of samples to generate per
+ * batch.
+ * @param num_samples_shape Shape of the 'num_samples' tensor.
+ * @param output Output tensor for the generated class indices.
+ * @param output_shape Shape of the 'output' tensor.
+ * @param with_replacement Boolean that determines whether a sampled class can appear more than once in the output.
+ * @param log_probs Boolean that determines whether to treat input probabilities as log probabilities.
+ * @param global_seed First seed value (key) of Phillox random number generation algorithm. (See RandomUniform for
+ * details)
+ * @param op_seed Second seed value (counter) of Phillox random number generation algorithm. (See RandomUniform for
+ * details)
+ */
+template <typename T, typename U, typename V>
+void multinomial(const T* probs,
+                 const Shape& probs_shape,
+                 const U* num_samples,
+                 const Shape& num_samples_shape,
+                 V* output,
+                 const Shape& output_shape,
+                 const bool with_replacement,
+                 const bool log_probs,
+                 const uint64_t global_seed,
+                 const uint64_t op_seed) {
+    auto total_inputs_elements_count = shape_size<Shape>(probs_shape);
+    auto total_output_elements_count = shape_size<Shape>(output_shape);
+    int test_nr = 1;
+
+    // If probabilities are log probabilities, exponentiate to get normal probabilities
+    std::vector<T> input_vals(total_inputs_elements_count);
+    if (log_probs) {
+        exp<T>(probs, input_vals.data(), total_inputs_elements_count);
```

Comment:
```suggestion
        exp(probs, input_vals.data(), total_inputs_elements_count);
```
T should be deduced from inputs. `<T>` use it only if compiler cannot deduce it correctly.

Reply:
tRemoved!